packages/utils/src/lib/report.ts

```diff
@@ -57,16 +57,26 @@ export function countCategoryAudits(
   refs: CategoryRef[],
   plugins: ScoredReport['plugins'],
 ): number {
+  // Create lookup object for groups within each plugin
+  const groupLookup = plugins.reduce<
+    Record<string, Record<string, ScoredReport['plugins'][0]['groups'][0]>>
+  >((lookup, plugin) => {
+    if (plugin.groups) {
+      lookup[plugin.slug] = plugin.groups.reduce<
+        Record<string, ScoredReport['plugins'][0]['groups'][0]>
+      >((groupLookup, group) => {
+        groupLookup[group.slug] = group;
+        return groupLookup;
+      }, {});
+    }
+    return lookup;
+  }, {});
```

Comment:
This code is mixing two paradigms, it would make sense to use one or the other, otherwise they negate the benefits of each other:

## 1 - Functional (declarative, immutable) paradigm

In my opinion, this should be favoured by default because the code is easier to understand and maintain (high-level, no side-effects, etc).

Using `reduce` is very common for functional code because it's a declarative way of aggregating an array into a single value. Although JavaScript cannot enforce it, mutating inside of `reduce` goes against the intended pattern. Specifically for accumulator, the contract is that you return a new value, instead of mutating it directly. The most common way to immutably create new values of objects (and arrays) is to use the spread operator:

```suggestion
  const groupLookup = plugins.reduce<
    Record<string, Record<string, ScoredReport['plugins'][0]['groups'][0]>>
  >((lookup, plugin) => {
    if (!plugin.groups.length) {
      return lookup;
    }
    return {
       ...lookup,
       [plugin.slug]: plugin.groups.reduce<Record<string, AuditGroup>>(
          (groupLookup, group) => ({ ...groupLookup, [group.slug]: group }),
          {},
       )
    };
  }, {});
```

## 2 - Imperative (low-level, mutable) paradigm

This is generally more verbose, harder to debug (intermediate states), etc. That makes often a worse choice for readability and maintenance, but it tends to have a slight performance advantage. So I would say the rule of thumb to only use when the code is known to be a performance bottleneck (e.g. runs extremely often, or handles a lot of data).

If this code would benefit from performance optimization (this should be based on some measurement, ideally), then the imperative approach could make sense. But in that case, there are more efficient ways to iterative over arrays than `.reduce`:

```suggestion
  const groupLookup: Record<string, Record<string, AuditGroup>> = {};
  for (const plugin of plugins) {
    if (!plugin.groups.length) {
      continue;
    }
    groupsLookup[plugin.slug] ??= {};
    for (const group of plugin.groups) {
       groupsLookup[plugin.slug][group.slug] = group;       
    }
  }
```

Reply:
Very nice and detailed feedback!! 

Thank you a LOT for your time! <3

Reply:
@matejchalk wow, that is very wide coverage of the issue, and one of the best comments I've ever seen! 😮 
Thank you for the description of both paradigms, you're right that they are mixed in my case. My fix will include declarative paradigm only, here and in another places (for this PR) ❤️